flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -0,0 +1,293 @@
+ */
+@Internal
+public class SinkTransformationTranslator<InputT, CommT, WriterStateT, GlobalCommT> implements
+		TransformationTranslator<Object, SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT>> {
+
+	@Override
+	public Collection<Integer> translateForBatch(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> transformation,
+			Context context) {
+
+		StreamGraphUtils.validateTransformationUid(context.getStreamGraph(), transformation);
+		final int parallelism = getParallelism(transformation, context);
+
+		int writerId = addWriter(
+				transformation,
+				parallelism, context);
+		int committerId = addCommitter(
+				writerId,
+				transformation,
+				new BatchCommitterOperatorFactory<>(transformation.getSink()),
+				parallelism, context);
+		addGlobalCommitter(
+				committerId >= 0 ? committerId : writerId,
+				transformation,
+				new BatchGlobalCommitterOperatorFactory<>(transformation.getSink()), context);
+		return Collections.emptyList();
+	}
+
+	@Override
+	public Collection<Integer> translateForStreaming(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> transformation,
+			Context context) {
+
+		StreamGraphUtils.validateTransformationUid(context.getStreamGraph(), transformation);
+
+		final int parallelism = getParallelism(transformation, context);
+
+		int writerId = addWriter(
+				transformation,
+				parallelism, context);
+		int committerId = addCommitter(
+				writerId,
+				transformation,
+				new StreamingCommitterOperatorFactory<>(transformation.getSink()),
+				parallelism, context);
+		addGlobalCommitter(
+				committerId >= 0 ? committerId : writerId,
+				transformation,
+				new GlobalStreamingCommitterOperatorFactory<>(transformation.getSink()), context);
+
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Add a sink writer node to the stream graph.
+	 *
+	 * @param sinkTransformation The transformation that the writer belongs to
+	 * @param parallelism The parallelism of the writer
+	 *
+	 * @return The stream node id of the writer
+	 */
+	private int addWriter(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			int parallelism,
+			Context context) {
+		final boolean hasState = sinkTransformation
+				.getSink()
+				.getWriterStateSerializer()
+				.isPresent();
+		checkState(sinkTransformation.getInputs().size() == 1);
+		@SuppressWarnings("unchecked")
+		final Transformation<InputT> input = (Transformation<InputT>) sinkTransformation
+				.getInputs()
+				.get(0);
+		final TypeInformation<InputT> inputTypeInfo = input.getOutputType();
+
+		final StreamOperatorFactory<CommT> writer =
+				hasState ? new StatefulWriterOperatorFactory<>(sinkTransformation.getSink()) : new StatelessWriterOperatorFactory<>(
+						sinkTransformation.getSink());
+
+		final int writerId = addOperatorToStreamGraph(
+				writer, input.getId(),
+				inputTypeInfo,
+				extractCommittableTypeInformation(sinkTransformation.getSink()),
```

Comment:
In general in the translation process here we make an implicit assumption that if we do not have a `Committer`, then we also do not have a `GlobalCommitter`, right? Because when we set the output type of the `Writer` we always set it to `CommT`. In addition, when we extract the output type of the writer in the `extractCommittableTypeInformation()` we only look if there is a `committableSerializer`. If this is not present then the `Writer` cannot communicate with the `GlobalCommitter`.

Is this something we want? And if yes, then we should have a validation step for the sink that throws an exception if we have a sink with a `GlobalCommitter` but without a `Committer`. If we do not want that, then I think we should adjust the `addWrier`. Is this analysis correct @guoweiM ?

Reply:
Hi @kl0u 
There might be some cases that only have `Writer` and `GlobalCommitter`.  And `addGlobalCommitter` would check whether the `CommittableTypeInformation` is `null`.

Hi, @aljoscha 
yes.